data/domain/player.tsx

```diff
@@ -1089,6 +1093,44 @@ export const updatePlayerTalentLevelExceptESBonus = (data: Map<string, any>) =>
         })
     }
 
+    // --- Skull of Major Talent (Grimoire 39) bonus ---
+    const grimoire = data.get("grimoire") as Grimoire | undefined;
+    const skullOfMajorTalentLevel = grimoire?.getUpgradeLevel?.(39) ?? 0;
```

Comment:
I think this is slightly too defensive in terms of handling of undefined. I don't want to say it's impossible but it's highly unlikely (and sign of bigger issues) if grimoire / tesseract / any domain isn't available at this stage.

Can you follow the existing pattern of fetch the domain at the top of the function and without the undefined?

Also the `getUpgradeLevel?` looks a bit odd to me, it's not possible for the function to not be defined is it? so these 2 changes will make the code a lot cleaner with your helpful `getUpgradeLevel` functions that always return a number.